File: typescript/base/src/main/vsdk/toolkit/environment/background/CubemapBackground.ts
import { ColorRgb } from "../../common/color/ColorRgb.js";
import { Vector3Dd } from "../../common/linealAlgebra/Vector3Dd.js";
import { RGBAImageUncompressed } from "../../media/RGBAImageUncompressed.js";
import { Ray } from "../geometry/element/Ray.js";
import { RayHit } from "../geometry/element/RayHit.js";
import { Box } from "../geometry/volume/Box.js";
import { Camera } from "../camera/Camera.js";
import { Background } from "./Background.js";

/** A six-image environment map sampled by casting a ray through a unit cube. */
export class CubemapBackground extends Background {
    private readonly backgroundImages: RGBAImageUncompressed[];
    private readonly boundingCube = new Box(1, 1, 1);

    public constructor(
        private camera: Camera,
        front: RGBAImageUncompressed,
        right: RGBAImageUncompressed,
        back: RGBAImageUncompressed,
        left: RGBAImageUncompressed,
        down: RGBAImageUncompressed,
        up: RGBAImageUncompressed,
    ) {
        super();
        this.backgroundImages = [front, right, back, left, down, up];
    }

    public override colorInDireccion(d: Vector3Dd): ColorRgb {
        const ray = new Ray(new Vector3Dd(), d.normalized());
        const hit = new RayHit();
        if (!this.boundingCube.doIntersectionFirstHit(ray, hit)) return new ColorRgb();

        const plane = this.classifyPlane(hit.n);
        let u = 1 - hit.u;
        let v = 1 - hit.v;
        let image: RGBAImageUncompressed;
        switch (plane) {
            case 1:
                image = this.backgroundImages[5]!;
                u = 1 - hit.v;
                v = hit.u;
                break;
            case 2:
                image = this.backgroundImages[4]!;
                u = hit.v;
                v = 1 - hit.u;
                break;
            case 3:
                image = this.backgroundImages[0]!;
                break;
            case 4:
                image = this.backgroundImages[2]!;
                break;
            case 5:
                image = this.backgroundImages[1]!;
                break;
            default:
                image = this.backgroundImages[3]!;
                break;
        }
        return image.getColorRgbBiLinear(u, v);
    }

    public getImages(): RGBAImageUncompressed[] {
        return this.backgroundImages;
    }

    public getCamera(): Camera {
        return this.camera;
    }

    public setCamera(camera: Camera): void {
        this.camera = camera;
    }

    private classifyPlane(normal: Vector3Dd): number {
        const x = Math.abs(normal.x());
        const y = Math.abs(normal.y());
        const z = Math.abs(normal.z());
        if (z >= x && z >= y) return normal.z() >= 0 ? 1 : 2;
        if (y >= x) return normal.y() >= 0 ? 3 : 4;
        return normal.x() >= 0 ? 5 : 6;
    }
}

File: java/base/src/main/vsdk/toolkit/environment/background/CubemapBackground.java
package vsdk.toolkit.environment.background;
import java.io.Serial;

import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.environment.geometry.element.Ray;
import vsdk.toolkit.environment.camera.Camera;
import vsdk.toolkit.environment.geometry.element.RayHit;
import vsdk.toolkit.environment.geometry.volume.Box;
import vsdk.toolkit.media.RGBAImageUncompressed;

public class CubemapBackground extends Background {
    @Serial private static final long serialVersionUID = 20060502L;

    private RGBAImageUncompressed [] images;
    private Camera camera;
    private Box boundingCube = null;

    public CubemapBackground(Camera camera,
                             RGBAImageUncompressed front,
                             RGBAImageUncompressed right,
                             RGBAImageUncompressed back,
                             RGBAImageUncompressed left,
                             RGBAImageUncompressed down,
                             RGBAImageUncompressed up) {
        super();

        this.camera = camera;
        images = new RGBAImageUncompressed[6];
        images[0] = front;
        images[1] = right;
        images[2] = back;
        images[3] = left;
        images[4] = down;
        images[5] = up;
        boundingCube = new Box(1, 1, 1);
    }

    /**
    @param d
    @return color as viewed in given direction
    */
    @Override
    public ColorRgb colorInDireccion(Vector3Dd d)
    {
        double u;
        double v;
        RGBAImageUncompressed img;

        d = d.normalized();
        Ray r = new Ray(new Vector3Dd(0, 0, 0), d);
        RayHit hit = new RayHit();
        if ( !boundingCube.doIntersectionFirstHit(r, hit) ) {
            return new ColorRgb();
        }
        int plane = classifyPlane(hit.normal);

        u = 1 - hit.u;
        v = 1 - hit.v;
        switch ( plane ) {
          case 1: // Top
            img = images[5];
            u = 1 - hit.v;
            v = hit.u;
            break;
          case 2: // Down
            img = images[4];
            u = hit.v;
            v = 1 - hit.u;
            break;
          case 3: // Front
            img = images[0];
            break;
          case 4: // Back
            img = images[2];
            break;
          case 5: // Right
            img = images[1];
            break;
          default: // Left
            img = images[3];
            break;
        }

        return img.getColorRgbBiLinear(u, v);
    }

    private int classifyPlane(Vector3Dd normal)
    {
        double ax = Math.abs(normal.x());
        double ay = Math.abs(normal.y());
        double az = Math.abs(normal.z());

        if ( az >= ax && az >= ay ) {
            return normal.z() >= 0 ? 1 : 2;
        }
        if ( ay >= ax ) {
            return normal.y() >= 0 ? 3 : 4;
        }
        return normal.x() >= 0 ? 5 : 6;
    }

    public RGBAImageUncompressed [] getImages()
    {
        return images;
    }

    public Camera getCamera()
    {
        return camera;
    }

    public void setCamera(Camera camera)
    {
        this.camera = camera;
    }
}
